gammaspec.py changes

KennellyT opened the pull request:
Functions are now undercase with underscore, added extra line on the bottom of gammaspec.py, and added a test to test spectrometry.py

KennellyT committed:
gammaspec.py has a new  function read_SPEC_ID_file that reads these files

KennellyT committed:
function names are undercase with underscore, added line to bottom of gammaspec.py and added test to test_spectometry.py

scopatz commented on tests/test_spectrometry.py:
```diff
@@ -91,7 +91,21 @@ def test_gross_counts():
     nc=sa.net_counts(gspec1, 475, 484, 1)
 
-
+def test_read_spec_id_file():
```
How is this working? You don't seem to actually call `read_spec_id_file()`

gonuke replied:
it's called at global scope

KennellyT commented:
@scopatz, It seems I forgot to add another gspec variable to call read_spec_id_file(). Ill edit this file to add gspec3 which calls read_spec_id_file().

KennellyT committed:
calls read_spec_id_file in test

scopatz commented:
OK! Just post to this issue when you have it figured out!

KennellyT committed:
gamma_effects and tests added

KennellyT commented:
gspec3 variable has been added to test_spectrometry.py and test_read_spec_id_file() has been edited to test the read_spec_id() function. I also added another module called gamma_effect.py that calculates the compton edge energy, backscatter energy ,single escape energy, and double escape energy when given an input energy. A test function has been created for this module and is called test_gamma_effects.py.
I will be drafting a tutorial on how to use gammaspec as well as gamma_effects to analyze .spe files within the week.

scopatz commented on pyne/gamma_effects.py:
```diff
@@ -0,0 +1,29 @@
+    """Returns compton edge energy from a given gamma energy"""  
+    E_compton = E- E/(1+(2*E/mec2))
+    return E_compton
+def backscatter(E):
```
PEP8: two spaces between functions.

scopatz commented on pyne/gamma_effects.py:
```diff
@@ -0,0 +1,29 @@
+
+"""
+
+mec2 = 511. #keV # Rest mass energy of an electron #.511 MeV 
```
global variables should be ALL_UPPERCASE

scopatz commented:
Thanks @KennellyT - just let me know here, when you think this PR is ready to go!

KennellyT committed:
Updated gamma_effects.py, gammaspec.py and test_spectrometry.py. Added new module srim_read.py with files Helium and test_srim_read.py

KennellyT commented:
Changed pull request to mirror comments. Also added srim_read.py, a function that reads srim files. Also added writecsv function in gammaspec.py.

KennellyT committed:
gammaspec.py write_csv function changed

KennellyT committed:
srim_read.py changed

KennellyT committed:
tests srim fixed

KennellyT committed:
spectanalysis new area functions and tests

KennellyT committed:
tests for spec updated

KennellyT committed:
new function gaussian_fit added

KennellyT committed:
spectanalysis.py edit

KennellyT committed:
took out gaussian test

KennellyT committed:
spectanalysis update

KennellyT commented:
Update of the above pull requests:
`gammaspec.py:`

`read_spec_id_file()`- reads and collects data from .spe files of the 'spec_id' format.
`write_csv()`-takes energy,counts, and channels from spectrum and writes them to csv file to be named by user.

`srim_read.py`
`read_srim_output()`-Reads a srim file

`spectanalysis.py`
`net_area()`-Calculates the net area under a peak by subtracting background from total
    counts under the peak. 
`end_point_average_area()`- Calculates the net area under a peak by end point averaging the 
    background counts and subtracting the background from total
    counts under the peak.

`gaussian_fit()`-Plots the Gaussian fit of the counts versus channel number. 

`fwhm()`- Calculates the full width half maximum of a peak.

scopatz commented on pyne/gamma_effects.py:
```diff
@@ -0,0 +1,35 @@
+"""Purpose:
```
We use numpydoc format.  So you don't need "purpose" and the following lines shouldn't be indented. thanks!

scopatz commented on pyne/gamma_effects.py:
```diff
@@ -0,0 +1,35 @@
+
+def compton_edge(E):
+    """Returns compton edge energy from a given gamma energy"""  
+    E_compton = E- E/(1+(2*E/MEC_2))
```
There should be a single space in front of the minus: `E - E/`

scopatz commented on pyne/gamma_effects.py:
```diff
@@ -0,0 +1,35 @@
+def single_escape(E):
+    """Returns single escape energy from a given gamma energy"""
+    if (E < 1022):	#keV
+        raise RuntimeError('E needs to be equal to or greater than 1022 keV to be valid')
```
This should be a ValueError instead of a RuntimeError, I would think.

scopatz commented on pyne/gamma_effects.py:
```diff
@@ -0,0 +1,35 @@
+def double_escape(E):
+    """Returns double escape energy from a given gamma energy"""
+    if (E < 1022):	#keV
+        raise RuntimeError('E needs to be equal to or greater than 1022 keV to be valid')
```
ValueError again.

scopatz commented on pyne/srim_read.py:
```diff
@@ -0,0 +1,165 @@
+                 long_straggs=None,long_stragg_units=None,lat_straggs=None,
+                 lat_stragg_units=None):
+        """Define srim specific variables """
+        super(srim, self).__init__()
```
You shouldn't need to call super, but it doesn't hurt.

scopatz commented on pyne/srim_read.py:
```diff
@@ -0,0 +1,165 @@
+warn(__name__ + " is not yet QA compliant.", QAWarning)
+
+
+class srim(object):
```
Class name should be `CapCase`, so this should be `Srim` or similar.

scopatz commented on pyne/srim_read.py:
```diff
@@ -0,0 +1,165 @@
+
+
+class srim(object):
+    """srim class includes srim specific variables"""
```
This docstring doesn't help users who don't know what srim is. Could you expand please? Thanks!

scopatz commented on pyne/srim_read.py:
```diff
@@ -0,0 +1,165 @@
+                 ion_units=None,proj_ranges=None,proj_range_units=None,
+                 long_straggs=None,long_stragg_units=None,lat_straggs=None,
+                 lat_stragg_units=None):
+        """Define srim specific variables """
```
This docstring should include the meaning and type of all of the arguments and keyword arguments (except `self`, which doesn't need to be documented.)

scopatz commented on pyne/srim_read.py:
```diff
@@ -0,0 +1,165 @@
+        return print_string
+
+def read_srim_output(filepath):
+    #def read_srim_output(file_path):
```
Please remove commented out code line

scopatz commented on tests/test_spectrometry.py:
```diff
@@ -2,16 +2,16 @@
 warnings.simplefilter("ignore", QAWarning)
 
-from pyne import gammaspec
+#from pyne import gammaspec
```
please remove commented out lines like this.

scopatz commented on tests/test_spectrometry.py:
```diff
@@ -91,8 +93,33 @@ def test_gross_counts():
+
+def test_end_point_average_area():
+    nc = sa.end_point_average_area(gspec1,475,484,var=5)
+    #assert_equal(nc,20355.5) 
```
Shouldn't the tests here have an assert? Is the problem that you are doing a float comparison? You should probably use `assert_almost_equals` in this case

scopatz commented:
Thanks for the update @KennellyT!

KennellyT committed:
changed gamma_effects, spectanalysis, srim_read,readme.rst,test_spectrometry after reviewing comments

KennellyT commented:
@scopatz- I took your comments and advice and implemented them into the pull request. readme.rst file had an issue where it states:

> conda install -c conda-forge pyne

> where VERSION should be replaced with the version number to be installed.

but version is not in the conda install -c statement nor is it needed to install pyne via conda via [https://anaconda.org/conda-forge/pyne](https://anaconda.org/conda-forge/pyne)

I have updated srim_read and class by adding the meaning of the variables. 
Thank you

KennellyT committed:
assert_almost_equal

KennellyT committed:
python 2 division issue with spectanalysis.py

KennellyT committed:
gammaspec epoly function, spectanalysis fwhm

KennellyT committed:
Added gv_format_spect2 to tests folder

KennellyT committed:
circleci did not accept printed equation that outputs from calc_energy_poly

KennellyT committed:
assert_almost_equal used for float

KennellyT committed:
new functions epoly_fit, resolution added

py1sl commented:
@KennellyT pleased to see this get some attention, its been on my todo list since I first wrote the initial read functions, when I was first learning python
It would be good to add peak find methods as well, i believe i wrote a mariscotti peak find routine for spectanalysis but didn't test it,  not needed for this PR

KennellyT commented:
@py1sl - I'll look at your mariscotti peak find routine/code as I have developed a peak finder function using the second derivative test and Savitzky Golay filter but I have not fully tested the function on a wide variety of gamma spectra.

katyhuff commented:
@KennellyT ... this seems ready to merge to me. 

I see that @py1sl has suggested a peak finding method as well. Perhaps we can merge this and add a future issue for peak finding. What do you think?

gonuke commented on pyne/spectanalysis.py:
```diff
@@ -146,8 +148,154 @@ def net_counts(spec, c1, c2, m):
+    y = spec.counts[c1:c2]
+    n = len(x)
+    mean = sum(x*y)/n                   
+    sigma = np.sqrt(sum(y*(x-mean)**2)/n)
```
Is this `sigma` ever used?

KennellyT replied:
So the sigma variable goes into the gauss function (line 240) `def gauss(x,amp,x_center,sigma):`but in hindsight, this operation can be done in a much cleaner and more efficient manner. I'll take out the gauss function (line 240) and replace it with either its own variable or make the function separate from the `gaussian_fit` function.

gonuke replied:
But this `sigma` does NOT get used in the `gauss()` function.... that function has an argument named `sigma` that gets populated automatically by the `curve_fit` method.  The initial guess for this value appears to be `n/2`.

KennellyT replied:
Ah, I now see the `sigma` issue. I'll replace the `n/2` with `sigma`. Thank you!

gonuke commented on pyne/spectanalysis.py:
```diff
@@ -146,8 +148,154 @@ def net_counts(spec, c1, c2, m):
+    return end_point_average
+
+def gaussian_fit(spec,c1,c2):
+    """Plots the gaussian fit to the raw data. 
```
It might make sense to separate the curve fitting from the plotting in case users want to fit without plotting. This would let you reuse it in the `fwhm()` function below.

KennellyT replied:
I concur and I will separate this function into two

gonuke commented on pyne/spectanalysis.py:
```diff
@@ -146,8 +148,154 @@ def net_counts(spec, c1, c2, m):
+    net_area = p - ((n/2.0)*(count1+count2))
+    return net_area
+
+def end_point_average_area(spec,c1,c2,var=5):
```
Should this be equivalent to `net_area` if `var` = 1.  I don’t think it is, thanks to python indexing. If it should be, then `net_area` should just Call this with `var=1`

gonuke reviewed (commented):
Thanks for this contribution @KennellyT - sorry that it has been sitting untended for so long...

I have a few little comments, but will otherwise look to merge this soon.

KennellyT committed:
sigma usage in gaussian_fit

KennellyT commented:
Appears that my additions are failing in `test_decay.py` when trying to download `http://data.pyne.io/origen-benchmark-hdf5-1.8.14.h5: E`. It seems that the link is down: `urllib.error.HTTPError: HTTP Error 404: Not Found` or `Unable to open/create file 'origen-benchmark-hdf5-1.8.14.h5'`

gonuke commented:
You might need to rebase your code against `develop`.  We've recently moved that file to a different location because data.pyne.io has gone down.  The default branch now gets this code from the github pyne/data repository

gonuke commented:
I'd like to push for a merge on this @KennellyT - let me know if you want any assistance in rebasing and moving this foward

ahnaf-tahmid-chowdhury committed:
Merge branch 'develop' into experimental_KennellyT

ahnaf-tahmid-chowdhury commented:
Hi @KennellyT, are you still working on this?
